substratest/assets.py

```diff
@@ -168,15 +139,54 @@ def load(cls, d):
             raise errors.TError(f"cannot parse asset `{d}`") from e
 
 
-class _Asset(_DataclassLoader, abc.ABC):
+class _InternalStruct(pydantic.BaseModel, _DataclassLoader, abc.ABC):
+    """Internal nested structure"""
+
+
+class _Asset(_InternalStruct, abc.ABC):
     """Represents assets stored in the Substra framework.
 
     Convert a dict with camel case fields to a Dataclass.
     """
 
 
-@dataclasses.dataclass(frozen=True)
-class Permission(_DataclassLoader):
+class _BaseFutureAsset(_Asset):
+    __session: typing.Any = None
+    _future_cls = None
+
+    def attach(self, session):
+        """Attach session to asset."""
+        object.__setattr__(self, '__session', session)
+        return self
+
+    @property
+    def _session(self):
+        # because self.__session doesn't work properly with Pydantic, we have to use
+        # __getattribute__ and __setattr__ (https://docs.python.org/3/reference/datamodel.html)
+        if not object.__getattribute__(self, '__session'):
+            raise errors.TError(f'No session attached with {self}')
+        return object.__getattribute__(self, '__session')
+
+    def future(self):
+        """Returns future from asset."""
+        return self._future_cls(self, object.__getattribute__(self, '__session'))
+
+
+class _FutureAsset(_BaseFutureAsset):
+    """Represents a single task that is executed on the platform."""
+    _future_cls = Future
+
+    def future(self):
+        assert hasattr(self, 'status')
+        assert hasattr(self, 'key')
+        return super().future()
```

Comment:
not related to this PR directly but could be removed.

Reply:
Should we do this in this PR instead of another one?

Reply:
It could be done later in a separate PR. :)

Reply:
Alright!